### PR TITLE
=str #22554 fix ClassCastException in SubSink

### LINKED ISSUE #22554: CCE in SubSink
It seems to fail in code that should handle a race condition when materialization is faster than the first `pullSubstream`.

```
[info] [ERROR] [03/14/2017 13:20:04.934] [akkaboom-akka.actor.default-dispatcher-3] [akka://akkaboom/user/StreamSupervisor-1/flow-2-0-mergePreferred] Error in stage [akka.http.impl.engine.client.PoolSlot$SlotProcessor@e9dfac6]: akka.stream.stage.GraphStageLogic$$anon$1 cannot be cast to scala.Function1
[info] java.lang.ClassCastException: akka.stream.stage.GraphStageLogic$$anon$1 cannot be cast to scala.Function1
[info]     at akka.stream.impl.fusing.SubSink.pullSubstream(StreamOfStreams.scala:600)
[info]     at akka.stream.stage.GraphStageLogic$SubSinkInlet.pull(GraphStage.scala:1053)
[info]     at akka.http.impl.engine.client.PoolSlot$SlotProcessor$$anon$1.establishConnectionFlow$1(PoolSlot.scala:192)
[info]     at akka.http.impl.engine.client.PoolSlot$SlotProcessor$$anon$1.onPush(PoolSlot.scala:205)
[info]     at akka.stream.impl.fusing.GraphInterpreter.processPush(GraphInterpreter.scala:500)
[info]     at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:402)
[info]     at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:570)
[info]     at akka.stream.impl.fusing.ActorGraphInterpreter$SimpleBoundaryEvent$class.execute(ActorGraphInterpreter.scala:41)
[info]     at akka.stream.impl.fusing.ActorGraphInterpreter$BatchingActorInputBoundary$OnNext.execute(ActorGraphInterpreter.scala:74)
[info]     at akka.stream.impl.fusing.GraphInterpreterShell.processEvent(ActorGraphInterpreter.scala:545)
[info]     at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:724)
[info]     at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:739)
[info]     at akka.actor.Actor$class.aroundReceive(Actor.scala:513)
[info]     at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:649)
[info]     at akka.actor.ActorCell.receiveMessage(ActorCell.scala:510)
[info]     at akka.actor.ActorCell.invoke(ActorCell.scala:479)
[info]     at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
[info]     at akka.dispatch.Mailbox.run(Mailbox.scala:224)
[info]     at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
[info]     at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
[info]     at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
[info]     at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
[info]     at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

### PR DESCRIPTION
**Pull request by jrudolph**
Fixes #22554.

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala [hunk 1]
```diff
@@ -590,23 +607,27 @@ final class SubSink[T](name: String, externalCallback: ActorSubscriberMessage 
...
 
-  def pullSubstream(): Unit = {
+  @tailrec
+  private def dispatchCommand(newState: CommandScheduledBeforeMaterialization): Unit =
     status.get match {
-      case f: AsyncCallback[Any] @unchecked ⇒ f.invoke(RequestOne)
-      case null ⇒
-        if (!status.compareAndSet(null, RequestOne))
-          status.get.asInstanceOf[Command ⇒ Unit](RequestOne)
+      case /* Materialized */ callback: AsyncCallback[Command] ⇒ callback.invoke(newState.command)
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala]: isn't the `@unchecked` still needed to silence the compiler?

**Comment by jrudolph** [comment]
Fixed a small bug that replaced all `cancel` calls with `request` semantics. `LazySourceSpec` failed, fortunately.

**Review by patriknw — APPROVED** [review]
LGTM

File: akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala [hunk 2]
```diff
@@ -590,23 +607,27 @@ final class SubSink[T](name: String, externalCallback: ActorSubscriberMessage 
...
+  def cancelSubstream(): Unit = dispatchCommand(CancelScheduledBeforeMaterialization)
 
-  def pullSubstream(): Unit = {
+  @tailrec
+  private def dispatchCommand(newState: CommandScheduledBeforeMaterialization): Unit =
     status.get match {
-      case f: AsyncCallback[Any] @unchecked ⇒ f.invoke(RequestOne)
-      case null ⇒
-        if (!status.compareAndSet(null, RequestOne))
-          status.get.asInstanceOf[Command ⇒ Unit](RequestOne)
```
Comment (jrudolph) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala]: The only thing that was really broken was that cast here and a few lines below. The rest is just cleanup.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala (hunk 1 above)
Comment (jrudolph) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala]: Yes, good point, will add it back.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala [hunk 3]
```diff
@@ -572,7 +572,24 @@ final class Split[T](val decision: Split.SplitDecision, val p: T ⇒ Boolean, va
...
+
+  // preallocated instances for both commands
+  /** A RequestOne command was scheduled before materialization */
+  case object RequestOneScheduledBeforeMaterialization extends CommandScheduledBeforeMaterialization(RequestOne)
+  /** A Cancel command was scheduled before materialization */
+  case object CancelScheduledBeforeMaterialization extends CommandScheduledBeforeMaterialization(Cancel)
+
+  /** Steady state: sink has been materialized, commands can be delivered through the callback */
+  // Represented in unwrapped form as AsyncCallback[Command] directly to prevent a level of indirection
+  // case class Materialized(callback: AsyncCallback[Command]) extends State
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala]: neat trick, I like it 👍

**Review by ktoso — APPROVED** [review]
LGTM

**Comment by jrudolph** [comment]
In theory it could happen in 2.4 as well, but the bug was in there forever and never happened. Should we still backport?

**Comment by patriknw** [comment]
@jrudolph  if it's an easy backport I think it would be good to do that

**Comment by jrudolph** [comment]
Seemed easy. Backport in #22567.

**Comment by johanandren** [comment]
Nice with the extra cleanup as well, I like!

**Pull request merged**